### PR TITLE
New model: HGATs

### PR DESCRIPTION
**Pull request by markzhao98**
## Types of changes

- [ ] Fix bugs
- [x] Add new feature
- [x] Update documentation

### CODE REVIEW

**Comment by you-n-g** [comment]
Please fix the CI error.

**Comment by you-n-g** [comment]
Could you please add a benchmark for Alpha360?

File: qlib/contrib/model/pytorch_hgats_ts.py [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+        for data, i in data_loader:
+
+            data = data.squeeze()
+            feature = data[:, :, 0:-1].to(self.device)
+            label = data[:, -1, -1].to(self.device)
+
+            GH = pd.get_dummies(big_G.values.squeeze()).values.astype("float64")
+            GH = GH[
+                big_G.index.isin(
+                    tsds.get_index()[i.tolist()].to_frame()["instrument"].values
```
Comment (you-n-g) [inline qlib/contrib/model/pytorch_hgats_ts.py:245]: Let's have a discussion of the content of big_G

File: qlib/contrib/model/pytorch_hgats_ts.py [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+                )
+            ]  # [~#stocks, #industries]
+
+            pred = self.GAT_model(feature.float(), GH)
```
Comment (you-n-g) [inline qlib/contrib/model/pytorch_hgats_ts.py:249]: The name should not be GAT anymore?

File: qlib/contrib/model/pytorch_hgats_ts.py [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+
+    def forward(self, x, GH):
+
+        out, _ = self.rnn(x)
+        hidden = out[:, -1, :]  # [~#stocks, #features]
+        att_weight = self.cal_attention(
+            hidden, torch.Tensor(GH).to(self.device)
+        )  # [~#stocks, #industries]
+
+        De = torch.Tensor(pinv(np.diag(GH.sum(axis=0)))).to(self.device)
```
Comment (you-n-g) [inline qlib/contrib/model/pytorch_hgats_ts.py:515]: The device should be managed in `class GATs(Model):`

File: qlib/contrib/model/pytorch_hgats_ts.py [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+    def __iter__(self):
+
+        for idx, count in zip(self.daily_index, self.daily_count):
+            yield np.arange(idx, idx + count)
+
+    def __len__(self):
+        return len(self.data_source)
+
+
+class GATs(Model):
```
Comment (you-n-g) [inline qlib/contrib/model/pytorch_hgats_ts.py:64]: rename it

File: qlib/contrib/model/pytorch_hgats_ts.py [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+            raise NotImplementedError(
+                "optimizer {} is not supported!".format(optimizer)
+            )
+
+        self.fitted = False
+        self.GAT_model.to(self.device)
+
+        self.bigG = pd.read_pickle("benchmarks/HGATs/hypergraph/CSI300.pkl")
+        self.num_ind = 29
+        # self.num_ind = (
```
[the remark this answers is not in the document]
Comment (markzhao98) [reply qlib/contrib/model/pytorch_hgats_ts.py:188]: done

File: qlib/contrib/model/pytorch_hgats_ts.py (hunk 4 above)
Comment (markzhao98) [reply qlib/contrib/model/pytorch_hgats_ts.py:64]: done

File: qlib/contrib/model/pytorch_hgats_ts.py (hunk 2 above)
Comment (markzhao98) [reply qlib/contrib/model/pytorch_hgats_ts.py:249]: done

File: qlib/contrib/model/pytorch_hgats_ts.py [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+            self.train_optimizer = optim.SGD(self.GAT_model.parameters(), lr=self.lr)
+        else:
+            raise NotImplementedError(
+                "optimizer {} is not supported!".format(optimizer)
+            )
+
+        self.fitted = False
+        self.GAT_model.to(self.device)
+
+        self.bigG = pd.read_pickle("benchmarks/HGATs/hypergraph/CSI300.pkl")
```
[the remark this answers is not in the document]
Comment (markzhao98) [reply qlib/contrib/model/pytorch_hgats_ts.py:186]: done. I've also included a "requirements.txt" file in the HGATs folder.

File: examples/benchmarks/HGATs/workflow_config_hgats_Alpha158.yaml [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    - class: DropnaLabel
+    - class: CSRankNorm
+      kwargs:
+        fields_group: label
+  label: ["Ref($close, -2) / Ref($close, -1) - 1"]
+port_analysis_config: &port_analysis_config
+  strategy:
+    class: TopkDropoutStrategy
+    module_path: qlib.contrib.strategy.strategy
+    kwargs:
```
[the remark this answers is not in the document]
Comment (markzhao98) [reply examples/benchmarks/HGATs/workflow_config_hgats_Alpha158.yaml:56]: Yes.

[the remark this answers is not in the document]
Comment (markzhao98) [reply examples/benchmarks/HGATs/workflow_config_hgats_Alpha158.yaml:56]: I've added the following lines to the yaml file and qrun still worked.
>       signal:
>         - <MODEL> 
>         - <DATASET>

File: qlib/contrib/model/pytorch_hgats_ts.py [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+            GH = pd.get_dummies(self.bigG.values.squeeze()).values.astype("float64")
+            GH = GH[
+                self.bigG.index.isin(
+                    dl_test.get_index()[i.tolist()].to_frame()["instrument"].values
+                )
+            ]  # [~#stocks, #industries]
+
+            if GH.shape[0] != 300:  # new securities added to the the CSI
+
+                break  # comment this out if you want to use interpolation for unknown industry information for the new securities
```
[the remark this answers is not in the document]
Comment (markzhao98) [reply qlib/contrib/model/pytorch_hgats_ts.py]: This is resolved by creating an 'unknown' industry for new securities being included to the CSI index.

**Comment by markzhao98** [comment]
The problem of the mis-interpretation of x_j should be resolved, together with the above mentioned issues.

File: qlib/contrib/model/pytorch_hgats_ts.py (hunk 3 above)
Comment (markzhao98) [reply qlib/contrib/model/pytorch_hgats_ts.py:511]: done

File: qlib/contrib/model/pytorch_hgats_ts.py (hunk 1 above)
Comment (markzhao98) [reply qlib/contrib/model/pytorch_hgats_ts.py:245]: This is resolved - the industry information is now drawn from the daily stock batch.

Comment (markzhao98) [reply qlib/contrib/model/pytorch_hgats_ts.py]: The problem with the mis-interpretation of x_j is also addressed.

File: examples/benchmarks/HGATs/README.md [hunk 9]
```diff
@@ -0,0 +1,4 @@
+# HGATs
```
[the remark this answers is not in the document]
Comment (markzhao98) [reply examples/benchmarks/HGATs/README.md:1]: News and Model Zoo list updated. Benchmark list pending Alpha360 benchmark.

File: qlib/contrib/model/pytorch_hgats_ts.py [hunk 10]
```diff
@@ -0,0 +1,547 @@
...
+        att_weight = self.softmax(attention_out)
+
+        return att_weight
+
+    def forward(self, x, GH):
+
+        out, _ = self.rnn(x)
+        hidden = out[:, -1, :]  # [#stocks, #features]
+
+        hidden_agg = torch.t(GH).mm(hidden)  # [#industries, #features]
```
Comment (you-n-g) [inline qlib/contrib/model/pytorch_hgats_ts.py:529]: The aggregation is a little weird. 
The scale is not aligned and the meaning of  `hidden_agg`  may be greatly different from hidden (But they will be transformed by the same `self.transformation` later)

**Comment by you-n-g** [comment]
Formating the code to fix the CI error is easy `python -m black .  -l 120`

File: examples/benchmarks/HGATs/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,4 @@
+pandas==1.1.2
+numpy==1.17.4
+scikit_learn==0.23.2
+torch==1.7.0
```
[the remark this answers is not in the document]
Comment (markzhao98) [reply examples/benchmarks/HGATs/requirements.txt:4]: done. Thanks for the check!

File: qlib/contrib/model/pytorch_hgats_ts.py (hunk 10 above)
Comment (markzhao98) [reply qlib/contrib/model/pytorch_hgats_ts.py:529]: The meaning of `hidden_agg` and `hidden` are different by design. `hidden_agg` corresponds to `xj` in the paper and `hidden` corresponds to `xi`. They go through the same linear transformation `P`.

Screen Shot 2021-12-10 at 11 30 56

File: docs/developer/code_standard.rst [hunk 12]
```diff
@@ -18,5 +18,5 @@ A common error is the mixed use of space and tab. You can fix the bug by inputin
 
 .. code-block:: python
 
-    pip install black
+    pip install black==21.6b0
```
Comment (you-n-g) [inline docs/developer/code_standard.rst:21]: If this `pip install --upgrade black`  works,  it will be better use it instead

**Comment by you-n-g** [comment]
@markzhao98 
I got this error.
Maybe this method is not supported in Pytroch 1.7.0 
https://pytorch.org/docs/1.7.0/search.html?q=nan_to_num&check_keywords=yes&area=default


image

**Pull request closed without merging**